src/mc-textopts.c

```diff
@@ -0,0 +1,305 @@
+            return false;                                                                                              \
+        }                                                                                                              \
+        has_##Name = true;                                                                                             \
+    ((void)0)
+
+#define END_IF_FIELD                                                                                                   \
+    continue;                                                                                                          \
+    }                                                                                                                  \
+    else((void)0)
+
+#define ERROR_PREFIX "Error parsing TextOpts: "
+
+bool mc_TextOptsPerIndex_parse(mc_TextOptsPerIndex_t *txio, bson_iter_t *iter, mongocrypt_status_t *status) {
+    *txio = (mc_TextOptsPerIndex_t){0};
+    txio->set = true;
+
+    bool has_strMaxLength = false, has_strMinQueryLength = false, has_strMaxQueryLength = false;
+    while (bson_iter_next(iter)) {
+        const char *field = bson_iter_key(iter);
+        BSON_ASSERT(field);
+
+        IF_FIELD(strMaxLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxLength' must be greater than zero");
+                return false;
+            }
+            txio->strMaxLength = OPT_I32(val);
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(strMinQueryLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMinQueryLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMinQueryLength' must be greater than zero");
+                return false;
+            }
+            txio->strMinQueryLength = val;
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(strMaxQueryLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxQueryLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxQueryLength' must be greater than zero");
+                return false;
+            }
+            txio->strMaxQueryLength = val;
+        }
+        END_IF_FIELD;
+
+        CLIENT_ERR(ERROR_PREFIX "Unrecognized field: '%s'", field);
+        return false;
+    }
+    return true;
+}
+
+bool mc_TextOpts_parse(mc_TextOpts_t *txo, const bson_t *in, mongocrypt_status_t *status) {
```

Comment:
Suggest adding tests similar to `test-mc-rangeopts.c` to exercise different parsing inputs and error scenarios: https://github.com/mongodb/libmongocrypt/commit/368036a2e0a0742ef7c7aaf1b8a8688aa93a0837 (can be copied from [this branch](https://github.com/mongodb/libmongocrypt/compare/master...kevinAlbs:libmongocrypt:text-search?expand=1))